[ENH, MRG] Add EpochsSpectrumArray and SpectrumArray classes

alexrockhill opened the pull request:
Makes API a lot nicer and consistent with MNE related to discussion here: https://mne.discourse.group/t/instantiating-epochsspectrum-and-spectrum-class-objects/7127.

Without the docstrings of the array, you actually have to look in `__set_state__` to figure out what is required in the `state` argument which is pretty confusing for me. The use case of ragged/unequal epochs is something I ran into too.

Okay with you @drammock ?

alexrockhill commented on mne/time_frequency/spectrum.py:
```diff
@@ -1179,6 +1179,56 @@ def __getitem__(self, item):
+    ----------
+    data : array, shape (n_channels, n_freqs)
+        The channels' power spectral density for each epoch.
+    %(freqs_tfr)s
+    %(info_not_none)s
```
```suggestion
    %(info_not_none)s
    %(freqs_tfr)s
```

alexrockhill commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1418,84 @@ def average(self, method="mean"):
+    ----------
+    data : array, shape (n_epochs, n_channels, n_freqs)
+        The channels' power spectral density for each epoch.
+    %(freqs_tfr)s
+    %(info_not_none)s
```
```suggestion
    %(info_not_none)s
    %(freqs_tfr)s
```

alexrockhill commented:
Looks like the failures are unrelated, good to go by me

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1418,84 @@ def average(self, method="mean"):
+    %(info_not_none)s
+    %(freqs_tfr)s
+    %(method_psd)s
+    events : None | array of int, shape (n_events, 3)
+        The events typically returned by the read_events function.
+        If some events don't match the events of interest as specified
+        by event_id, they will be marked as 'IGNORED' in the drop log.
+        If None (default), all event values are set to 1 and event time-samples
+        are set to range(n_epochs).
+    event_id : int | list of int | dict | None
+        The id of the event to consider. If dict,
+        the keys can later be used to access associated events. Example:
+        dict(auditory=1, visual=3). If int, a dict will be created with
+        the id as string. If a list, all events with the IDs specified
+        in the list are used. If None, all events will be used with
+        and a dict is created with string integer names corresponding
+        to the event id integers.
+    metadata : instance of pandas.DataFrame | None
+        See :class:`mne.Epochs` docstring for details.
+    %(selection)s
+    %(drop_log)s
```
wondering what makes the most sense for these.  Would like opinions of other devs (@hoechenberger @larsoner @mscheltienne @agramfort) as well as @alexrockhill... Off the top of my head:

- `events` I can see why folks might want to pass this in but is there any reason it needs to be a "proper" events array (N x 3)?  Seems like only the last column is going to matter? it *can* be set after-the-fact so we *could* omit it from the constructor, but my gut is that this will be frequently used so maybe best to leave it here?
- `event_id` again I can see why folks might want this (for `__getitem__`)... it can be set after-the-fact and probably doesn't hurt to allow that; not sure about *forcing* it to be after-the-fact if we're going to have `events` as a param though. Thoughts?
- `metadata` can be set after instantiation so I'm inclined to leave it out of the constructor
- `selection` seem like this shouldn't be needed if you're working with `np.array` data. Is there a use case you're thinking of where users would want to pass anything besides `selection=np.arange(data.shape[0])`? If not we can just set that internally
- `drop_log` similar to `selection`; is there a use case you're thinking of where users would want to pass anything besides `drop_log=tuple(tuple() for _ in range(data.shape[0]))`? If not we can just set that internally

If we do end up keeping `events` and `event_id` as-is, the docstring should be deduped with `EpochsArray` (where it seems you copy-pasted these entries from)

alexrockhill replied:
+1 on simplifying the constructor, as inferred, this was copy-pasted with the thought that it should just be similar to `EpochsArray` and without much more thought than that. I think `events` and `event_id` are helpful in the constructor so people don't lose expected functionality.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1179,6 +1179,56 @@ def __getitem__(self, item):
+    See Also
+    --------
+    mne.create_info
+    mne.EpochsArray
+    mne.EvokedArray
+    mne.io.RawArray
+    EpochsSpectrumArray
```
```suggestion
    mne.EvokedArray
    mne.io.RawArray
    EpochsSpectrumArray
    Spectrum
```
logic here is to only crossref classes with one degree of difference (i.e., EpochsArray is not linked because it is 2 degrees different from SpectrumArray: epoched & not spectral)

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1418,84 @@ def average(self, method="mean"):
+    --------
+    mne.create_info
+    mne.EpochsArray
+    mne.EvokedArray
+    mne.io.RawArray
```
```suggestion
    EpochsSpectrum
```
see prior comment for proposed logic of which ones to crossref.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1179,6 +1179,56 @@ def __getitem__(self, item):
+                sfreq=info["sfreq"],
+                dims=("channel", "freq"),
+                freqs=freqs,
+                inst_type_str="Raw",
+                data_type="Average Power Spectrum",
```
If the data comes from an array we don't know if it was computed on Raw or averaged (Evoked) data. Also we don't know if it's power/amplitude, real/complex, etc.

```suggestion
                inst_type_str="Array",
                data_type="Unknown",
```

This will probably break some tests (assuming we're testing thoroughly) so hopefully those test failures can guide what needs to change elsewhere

alexrockhill replied:
So real/complex can be inferred from the data type. We ask specifically for power in the constructor so maybe we should require that let the `amplitude` kwarg in `Spectrum.plot` handle that? That'd be my 2 cents but not highly opinionated, whatever is simplest

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1418,84 @@ def average(self, method="mean"):
+                sfreq=info["sfreq"],
+                dims=("epoch", "channel", "freq"),
+                freqs=freqs,
+                inst_type_str="Epochs",
+                data_type="Epochs Power Spectra",
```
As before with SpectrumArray

```suggestion
                inst_type_str="Array",
                data_type="Unknown",
```

drammock commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -386,3 +387,31 @@ def test_spectrum_kwarg_triaging(raw):
+    spect2 = SpectrumArray(
+        data=spect.get_data(), freqs=spect.freqs, info=spect.info, method=spect.method
+    )
+    assert_array_equal(spect.get_data(), spect2.get_data())
+    assert_array_equal(spect.freqs, spect2.freqs)
```
should be possible to just do
```
    assert spect == spect2
```
because we define `__eq__` on these... though perhaps you're right that they won't/can't/needn't pass that stricter test?

alexrockhill replied:
I tried and it failed, I think because you don't pass the provenance in the constructor, that's going to make them not identical

drammock commented on tutorials/simulation/10_array_objs.py:
```diff
@@ -224,39 +224,7 @@
-
-
-spectrum = spectrum_from_array(data=psd_ave, freqs=freqs, inst_info=info)
+spectrum = mne.time_frequency.spectrum.SpectrumArray(
```
Need to update the text in the preceding paragraph about how this is done / its experimental status. Also:

```suggestion
spectrum = mne.time_frequency.SpectrumArray(
```

drammock reviewed (commented):
I pushed a commit with a couple changes too far outside the diff to do as web-interface suggestions. See my remaining comments below.

alexrockhill commented on mne/time_frequency/spectrum.py:
```diff
@@ -1179,6 +1179,56 @@ def __getitem__(self, item):
+        The channels' power spectral density for each epoch.
+    %(info_not_none)s
+    %(freqs_tfr)s
+    %(method_psd)s
```
Sounds reasonable, I haven't looked in depth but from a cursory glance, it seems `method` is used when the tapers are returned or the welch segments are unaggregated which I don't think we need to support right away. If someone asks for it, not too hard to add as a followup. Also give this some time in `dev` to make sure no one runs into issues before making it more complicated.

alexrockhill commented on mne/time_frequency/spectrum.py:
```diff
@@ -1179,6 +1179,56 @@ def __getitem__(self, item):
+    data : array, shape (n_channels, n_freqs)
+        The channels' power spectral density for each epoch.
+    %(info_not_none)s
+    %(freqs_tfr)s
```
I took out method but `mne.time_frequency.psd_array_welch/multitaper` return `freqs` and maybe it's a bit paranoid but adding a round trip from `freqs` to `fmin`/`fmax` and back to `freqs` one more place for sampling rate or what have you to get mixed up. Also, I cut out power noise so the `freqs` are actually modified and that wouldn't work for my use case. If that's not convincing, I'm fine doing `fmin`, `fmax` and then modifying the `psd` object for my use case if it's important to you that the API matches.

drammock replied:
> I cut out power noise so the `freqs` are actually modified and that wouldn't work for my use case

So you mean you have bins that are missing / your bins are not equally spaced?  I guess that's a compelling use case.  And since folks aren't really supposed to use the `(Epochs)Spectrum` constructors directly, I'm OK to sacrifice API similarity in this case.

alexrockhill commented:
Okay, I think the outstanding things here are:

- [ ] Decide on `fmin`/`fmax` or `freqs` constructor
- [ ] Think about if we want to support passing amplitude with a kwarg (`psd_array_welch/multitaper` only supports output `power` or `complex` so I would vote just supporting those which can be inferred from data type)
- [ ] Decide about what Epochs-level data to put in the `EpochsSpectrumArray` API (I think it's a good compromise now but happy to hear from all the opinions @drammock requested to weigh in)

Otherwise looks good, thanks for the review @drammock.

Oh one more thing, I set `Spectrum._inst_type` to `np.ndarray` for the arrays, it's not used anywhere yet so this won't break anything but I'm not sure what the plan for that attribute was so just wanted to make sure that's the right move for future compatibility.

drammock commented on tutorials/simulation/10_array_objs.py:
```diff
@@ -212,9 +212,8 @@
-# This is slightly
-# experimental at the moment but works. An API for doing this may follow.
+# spectrum using the :class:`~mne.time_frequency.SpectrumArray` class.
+# This is slightly experimental at the moment but works.
```
Let's assume that by the end of this PR it won't be experimental any more.  Otherwise we'll probably forget to update this before release.

```suggestion
```

drammock reviewed (commented):
OK looking pretty good, thanks for tackling this!  Left a couple more comments/questions, let's wait and see what other devs say now.

alexrockhill commented on mne/time_frequency/spectrum.py:
```diff
@@ -1429,32 +1427,15 @@ class EpochsSpectrumArray(EpochsSpectrum):
-    %(drop_log)s
+    %(events_epochs)s
+    %(event_id)s
+    %(metadata_epochs)s
```
I left it in, I could see a use case for it here but that sounds reasonable to modularize if we prefer they set metadata later, I'll take it out

alexrockhill commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -393,8 +393,11 @@ def test_spectrumarray_raw(raw):
+        freqs=spect.freqs,
+        info=spect.info,
     )
+    spect2.plot()
```
Plotting is when I caught a lot of the errors so I thought it would be helpful to add it to the test. I'll add a separate test.

alexrockhill commented:
Okay to merge? The `EpochsSpectrumArray` constructor is minimal so we can always follow up with a PR adding more things to the constructor if necessary. This seems like a good start.

agramfort commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -386,3 +387,55 @@ def test_spectrum_kwarg_triaging(raw):
     raw.plot_psd(ax=axes)
+
+
+def test_spectrumarray_raw(raw):
```
```suggestion
def test_spectrum_array_raw(raw):
```

agramfort commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -386,3 +387,55 @@ def test_spectrum_kwarg_triaging(raw):
+    assert_array_equal(spect.freqs, spect2.freqs)
+
+
+def test_epochsspectrumaray(epochs):
```
```suggestion
def test_spectrum_array_epochs(epochs):
```

agramfort commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -386,3 +387,55 @@ def test_spectrum_kwarg_triaging(raw):
+    assert_array_equal(spect.freqs, spect2.freqs)
+
+
+def test_plot_spectrumaray(raw, epochs):
```
```suggestion
def test_plot_spectrum_array(raw, epochs):
```

agramfort commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -386,3 +387,55 @@ def test_spectrum_kwarg_triaging(raw):
+            info=spect_orig.info,
+        )
+    )
+    spect_orig = epochs.compute_psd()
```
I would test the plotting in the test functions above to avoid the recomputation of the psd which is not free or use pytest fixture

drammock replied:
opened #11811

agramfort commented:
Ok for me

drammock commented:
> Okay to merge?

Your checklist [above](https://github.com/mne-tools/mne-python/pull/11803#issuecomment-1636489783) has only 1 of 3 items checked off (plus there's a fourth item that doesn't have a checkbox --- the question about `_inst_type` --- which is also not resolved yet).  So no: not okay to merge.  Is there some rush on your end?

alexrockhill commented:
Not a huge rush, I'm sharing a script that uses it with someone else so it's a bit of a pain to describe how to install a feature branch.

I was going to let whoever merged the PR check those boxes to make sure they're okay with them, it's fine by me how it is right now (after I implement the latest review).

alexrockhill commented:
I just went ahead and added the fixtures so everyone wins :smile:

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1415,61 @@ def average(self, method="mean"):
+
+    Notes
+    -----
+        .. versionadded:: 1.5.0
```
```suggestion
        .. versionadded:: 1.5
```

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1368,6 +1415,61 @@ def average(self, method="mean"):
+                dims=("epoch", "channel", "freq"),
+                freqs=freqs,
+                inst_type_str="Array",
+                data_type="Unknown",
```
OK I'm convinced by your argument @alexrockhill that we can infer `data_type` from the `dtype`... It will make the `__repr__` much nicer.  I don't think we should say "Fourier coefficients" or "(Averaged) Power spectrum" though; I'm thinking maybe "Complex spectrum" or "Real spectrum". I do think that `method` should be "unknown" though, so we would end up with

```
<Real spectrum (from Array, unknown method) | {dims}, {freq_range}>
<Complex spectrum (from Array, unknown method) | {dims}, {freq_range}>
```

alexrockhill replied:
Sounds good, looks like `Power Spectrum` is used elsewhere so I put that. Reading your comment more closely, you don't think power spectrum and prefer real though. I just think that might disambiguate as a double check for anyone providing amplitude that in fact it is supposed to be power but I can change to `Real` if you prefer.

drammock replied:
> I don't think we should say "Fourier coefficients" or "(Averaged) Power spectrum"

I thought this was pretty clear, but since you didn't do this I'll spell out my reasoning: I don't think that seeing "power spectrum" in the repr is going to be effective at reminding people that they should only pass power (not amplitude) values to the constructor. I also don't see the point of (implicitly) restricting the input to be power, since there's effectively no way we can check/verify that it's power. I'd much rather make the documentation (including repr) be agnostic about what is being fed in, and make downstream methods capable of not doing the wrong thing (i.e., the plotting function should know when to put real units vs unknown/arbitrary units; users can always overwrite the plot axis labels if they know what form their data is).

By the way in local testing the repr raises an error; I'll push a commit to fix in a minute or two.

alexrockhill replied:
That's fine with me to make it accept accept more general inputs i.e. amplitude, I just thought it was simpler to require people to use power since they can be easily interconverted and if that was the case to remind them in case they didn't see it in the constructor. I think that might be slightly easier to maintain although less general but again I'm not strongly opinionated about it, the real/complex is okay.

drammock replied:
I've fixed the REPR but after looking at what it would take to make the plotting functions agnostic to power vs amplitude I'm actually going to backpedal here; it's too much work.  I'll add a docstring note caveat that we assume power.

drammock reviewed (changes requested):
requesting changes to block merge until REPR issues are fixed

drammock commented:
playing around locally I'm getting a lot of errors. Things like this don't work:

```py
import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = (sample_data_folder / 'MEG' / 'sample' /
                        'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False, preload=False)
events = mne.find_events(raw, stim_channel='STI 014')
event_dict = {'auditory/left': 1, 'auditory/right': 2}
epochs = mne.Epochs(raw, events, event_id=event_dict, preload=False,
                    proj=False)

epo_spectrum = epochs.compute_psd(output="complex")
arr_spectrum = mne.time_frequency.EpochsSpectrumArray(
    epo_spectrum.get_data(),
    epo_spectrum.info,
    epo_spectrum.freqs
)
arr_spectrum.get_data()  # <-- fails with IndexError
```



Clearly tests are not thorough enough, marking as draft.

alexrockhill commented:
Are you getting any more errors? That error is because the tapers are returned when `output='complex'`. I think maybe we shouldn't support that actually because the multitapers are of the order `(n_epochs, n_channels, n_tapers, n_freqs)` and the welch segments are of the order `(n_epochs, n_channels, n_freqs, n_segments)`. I think the two options are, 1) require `method` be passed and that it be `welch` or `multitaper` or 2) don't support complex and > 3D data. What do you think @drammock?

alexrockhill commented:
Ok, I:

- Tested welch and multitaper unaggregated and power
- Implemented thorough checks that the data was the right shape
- Fixed bugs that came up when different data types were plotted

I think it's very thorough now and I'm pretty happy with it

drammock commented:
> I think the two options are, 1) require `method` be passed and that it be `welch` or `multitaper` or 2) don't support complex and > 3D data. What do you think @drammock?

hmmm.  I think I'd like to eventually support people passing 3D data for SpectrumArray and 4D data for EpochsSpectrumArray. But maybe for now it's enough to support the lower-dimensional use cases and wait and see if folks ask for the higher-dimensional ones. Which, by exension, means I think we don't need to support complex data (yet) either.  So yeah, let's restrict to real 2D (no epochs) or 3D (epochs) data for now. Do you have time to finish this before the 1.5 release, or should I take over?

alexrockhill commented:
I think it's finished and I added pretty comprehensive support for complex cases already. It took a lot longer but it helped find a lot of bugs in plotting.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1192,6 +1232,10 @@ class SpectrumArray(Spectrum):
     %(info_not_none)s
     %(freqs_tfr)s
+    %(method_psd)s
+    weights : ndarray
+        The weights used for averaging across tapers. Only required
+        for :func:`mne.time_frequency.psd_array_multitaper` with ``output='complex'``.
```
per my most recent comment, let's assume YAGNI on supporting complex data with `(Epochs)SpectrumArray`, so this param can go away.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1454,26 +1526,49 @@ def __init__(
+        if events is not None:
+            _check_data_shape("epoch", 0, data, events.shape[0])
+        _check_data_shape("channel", 1, data, _pick_data_channels(info).size)
+        if data.ndim == 4:
+            _check_option("method", method, ("welch", "multitaper"))
+            dims = (
+                ("epoch", "channel", "taper", "freq")
+                if method == "multitaper"
+                else ("epoch", "channel", "freq", "segment")
+            )
+        if np.iscomplexobj(data) and method == "multitaper":
+            actual = None if weights is None else weights.size
+            expected = data.shape[list(dims).index("taper")]
+            if actual != expected:
+                raise ValueError(
+                    f"Expected weights to be {expected} size, got {actual}"
+                )
+        else:
+            _check_option("weights", weights, [None])
```
ditto

drammock commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -389,49 +386,107 @@ def test_spectrum_kwarg_triaging(raw):
-    )
-    assert_array_equal(epochs_psd.get_data(), epochs_psd2.get_data())
-    assert_array_equal(epochs_psd.freqs, epochs_psd2.freqs)
+def _get_psd_varients(raw_psds, epochs_psds):
```
```suggestion
def _get_psd_variants(raw_psds, epochs_psds):
```

drammock commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -389,49 +386,107 @@ def test_spectrum_kwarg_triaging(raw):
-    """Test SpectrumArray and EpochsSpectrumArray plotting."""
-    psds = list()
-    psds.append(
+def test_spectrum_arrays(raw_psds, epochs_psds, tmp_path):
```
what is the motivation / rationale for combining these into one monolithic test?  IMO test functions are easier to write and read when they do *less* stuff, not *more*. (asking for reasons here, not necessarily changes)

drammock replied:
OK. If they really are that similar, the preferred way to deduplicate in tests is to either (1) factor out a helper function, or  (2) use a `@pytest.mark.parametrize` decorator. Running for-loops inside tests is not good because when the test fails it's not obvious which iteration of the loop you were on when it failed, which makes debugging harder.

drammock replied:
`git grep "def _get_inst"` will show you the way --- `request.getfixturevalue`

drammock commented:
> I added pretty comprehensive support for complex cases already. 

Yeah, I was just realizing that as I reviewed the changes

> It took a lot longer but it helped find a lot of bugs in plotting.

Bugs that affect the existing Spectrum classes?  Or that only come up for the Array classes?

> I think it's finished

Sounds like you don't want to work on this any further. If that's correct, then FYI it's pretty frustrating when you say things like "[What do you think @drammock?](https://github.com/mne-tools/mne-python/pull/11803#issuecomment-1652603399)" and then make up your own mind and do a bunch of work without waiting for the input you requested, then aren't receptive when the requested input arrives.

If you don't want to do any more work here LMK, I'm OK with taking over this PR.

alexrockhill commented:
I personally am not using complex input to an `EpochsSpectrumArray` but since it's output for the psd functions, I could see it as a use case since the support for complex was added to those methods.

There's a lot of back-and-forth, agreed but here's the main point and reasoning: I originally planned to do just power for the sake of time (not implemented yet) but then I didn't hear back and thought you would be +1 for supporting complex numbers so I went ahead and implemented that. Now that it's implemented, it doesn't seem like too much of a maintenance burden, just a few more pytest error checks and covers the complete case of `Spectrum` objects you could get not through the array method. So my vote is to keep it.

alexrockhill commented:
Just thinking about this, `metadata` is in the `EpochsArray` constructor so might be nice to have it in the `EpochsSpectrumArray` constructor for consistency. I know you can set it after and we've discussed this but it would be nice if things were consistent.

alexrockhill commented:
What's the status on this PR?

I could see a use case for unaggregated complex spectra in the case you want to do some kind of phase analysis (e.g. phase-amplitude coupling) and don't have even length epochs.

If that's not convincing, then let's just take out complex support.

drammock commented:
sorry, I got captured by a grant deadline, it was on my list to come back to this yesterday but I ran out of time. Will prioritize for today.

alexrockhill commented:
No worries, next week is great too, just didn't want it to get forgotten.

drammock commented:
we chatted about this PR in the dev meeting today, to gauge opinions on whether supporting complex data as input to `SpectrumArray` and `EpochsSpectrumArray` is important or YAGNI. General feeling was:

- PRO: it would be nice to include it for symmetry reasons, since in some cases our `.compute_psd()` methods return complex data.
- CON 1: it makes it hard to validate the input data (specifically the dimensionality). Currently we *output*:
    - 2D real-valued `Spectrum` (welch, multitaper on Raw/Evoked)
    - 3D real-valued `Spectrum` (unaggregated welch on Raw/Evoked)
    - 3D complex-valued `Spectrum` (unaggregated multitaper on Raw/Evoked)
    - 3D real-valued `EpochsSpectrum` (welch, multitaper on Epochs)
    - 4D real-valued `EpochsSpectrum` (unaggregated welch on Epochs)
    - 4D complex-valued `EpochsSpectrum` (unaggregated multitaper on Epochs)
  If we only accept real-valued input, we can safely only support 2 use cases: 2D real data -> `Spectrum`, 3D real data -> `EpochsSpectrum`. In contrast, if we allow complex input, for example how do we decide if a 3D complex input should be treated as *unaggregated estimates from continuous data* versus *normal estimates from epoched data*?
- CON 2: supporting complex data means thinking a lot about downstream processing effects (i.e., does it make sense to plot? if so, is `abs()` a sufficient approach for converting to real-valued numbers? etc) which means significant maintenance burden. Now, we already kinda have to do this for the unaggregated multitaper output, but importantly that is the *only* case where we allow complex data so (1) we know what is the sensible way to convert to real-valued when needed (i.e. aggregating with precomputed multitaper weights) and (2) we can rely on knowing the dimension of the data (because we created the object).

All this has made me realize something that I previously overlooked / didn't discuss in the meeting, which is that when we originally added complex support, *we were outputting numpy arrays* so the ramifications didn't need to be traced through the rest of our API to make sure the complexness of the data didn't break things. When I created the Spectrum classes, support for complex unaggregated multitaper output was included for legacy reasons, but it probably would have been better to not support it in the classes (and to keep open a separate array-yielding code path for users who want the unaggregated multitaper output).

Which leads me to ask @mmagnuski (who originally requested the complex unaggregated multitaper output) a few questions:

1. what was your original use case that led to that request?
2. do you know others who use MNE's unaggregated multitaper estimates in their work?
3. do you foresee a use case for computing complex-valued spectral estimates outside of MNE and then putting them into a `Spectrum` or `EpochsSpectrum` object?

alexrockhill commented:
Those are pretty reasonable concerns: you have to pass the method if it's complex and I added checks that the dimensions matched the size of the frequencies passed so I think it the checks for the wrong data make it very safe (i.e. even if frequencies and `n_segments`/`n_tapers`  are same size, the method is required so that would disambiguate the dimensions. I really can't think of a way you could pass the wrong data).

mmagnuski commented:
@drammock 
Hi, I am fine with NOT supporting complex per-taper output in Spectrum classes. The reason for returning complex per-taper output is mostly for calculating multitaper coherence (and other similar measures) - this is done per taper and then averaged. [This](https://www.nature.com/articles/s41598-020-74057-1) is an example paper that showcases something similar.
But in general I don't think it would be used frequently and it does not make much sense to support it in the Spectrum object - that might be more time spent implementing and testing than actual user-time spent using the code. :)

agramfort commented:
what's the status here? I see some remaining comments by @drammock. It's a matter of agreeing if we support complex or not?

alexrockhill commented:
Those are pretty reasonable concerns: you have to pass the method if it's complex and I added checks that the dimensions matched the size of the frequencies passed so I think it the checks for the wrong data make it very safe (i.e. even if frequencies and `n_segments`/`n_tapers`  are same size, the method is required so that would disambiguate the dimensions. I really can't think of a way you could pass the wrong data). 

> what's the status here? I see some remaining comments by @drammock. It's a matter of agreeing if we support complex or not?

Yes, that's my understanding we should agree whether to support complex psds.

One last point: I think from a maintenance perspective having complex spectrumarrays was actually very helpful because in testing complete cases, there were several bugs in how plotting was handled.

drammock commented:
I'll restate her the pros and cons from [my comment above](https://github.com/mne-tools/mne-python/pull/11803#issuecomment-1693635154):

- PRO: it would be nice to include it for symmetry reasons, since in some cases our .compute_psd() methods return complex data.
- CON 1: it makes it hard to validate the input data (specifically the dimensionality).
- CON 2: supporting complex data means thinking a lot about downstream processing effects.

@mmagnuski's comment suggests to me that we should probably deprecate complex output support in the `(Epochs)Spectrum` classes, and provide an alternate code path that outputs NumPy arrays, for folks like @mmagnuski who really want access to the individual taper estimates.  If we do that, then the main "pro" of allowing complex input to the `(Epochs)SpectrumArray` classes goes away. I'll gladly acknowledge that some issues with plotting were revealed by the effort of adding that support, but IMO that does not constitute a "maintenance win" --- we can easily keep the plotting bugfixes while eliminating the complex support, and keeping complex support is still more lines of code and more branching code paths and more downstream effects to keep track of, all for the sake of (as far as I've heard) no known real-world use case for why it's preferable to have the complex data in a `Spectrum`-like container.

My vote here is:

1. a PR that deprecates complex output in the Spectrum classes
2. a PR that adds a spectrum pytest fixture *and adds it to all the existing tests that use Spectrum objects* (which was not done here, and is why I initially requested a separate PR for that in the first place)
3. a PR that adds the new `(Epochs)SpectrumArray` classes with only real-valued input allowed (and no support for unaggregated-welch-style input, i.e., 2D is always `SpectrumArray`, 3D is always `EpochsSpectrumArray`), and associated tests

alexrockhill commented:
That's fine with me, I'll remove complex support, you spent a lot of time and effort making spectrum classes @drammock so I'm happy to differ to your judgment and I agree there are not a lot of apparent use cases that come to mind--phase doesn't make any sense, it's not time-frequency, thanks for the direct communication, that helps move forward.

alexrockhill commented:
Ok I removed complex support and this is ready for review. I changed the `data_type` to `Power Spectrum` from `Real`/`Complex Spectrum` because of the change not to support complex input. It's just for the display, I really don't care that much so just let me know if anyone wants it back to `Real` or whatever else.

drammock commented on mne/conftest.py:
```diff
@@ -298,9 +298,9 @@ def raw_ctf():
+def raw_spectrum(raw):
     """Get raw with power spectral density computed from mne.io.tests.data."""
-    return [raw.compute_psd(method=method) for method in ("welch", "multitaper")]
+    return raw.compute_psd()
```
I don't think we have a good reason to test both Welch and multitaper in the fixture; the differences aren't relevant to most tests (and should be covered adequately by the tests of the welch/mutitaper array methods used under the hood)

drammock commented on mne/conftest.py:
```diff
@@ -298,9 +298,9 @@ def raw_ctf():
 
 @pytest.fixture(scope="function")
-def raw_psds(raw):
+def raw_spectrum(raw):
```
for naming consistency

drammock commented on mne/time_frequency/__init__.py:
```diff
@@ -49,14 +55,3 @@
-    read_spectrum,
-)
-from ._stft import stft, istft, stftfreq
-from ._stockwell import tfr_stockwell, tfr_array_stockwell
```
when you merged in main, this wasn't properly integrated into the new lazy loading scheme

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -429,7 +429,7 @@ def _check_values(self):
         assert self._data.shape == self._shape
         # negative values OK if the spectrum is really fourier coefficients
-        if np.iscomplexobj(self._data):
+        if "taper" in self._dims:
```
reverting to reduce churn, esp. given that we now plan to deprecate support for complex data anyway.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -768,23 +766,10 @@ def plot_topo(
-            seg_axis = self._dims.index("segment")
-            psds = np.nanmedian(psds, axis=seg_axis)
-        if np.iscomplexobj(psds):  # convert to power for plotting
-            psds = (psds * psds.conj()).real
```
reverting; will not be needed.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -901,17 +883,6 @@ def plot_topomap(
-            seg_axis = self._dims.index("segment")
-            psds = np.nanmedian(psds, axis=seg_axis)
-        if np.iscomplexobj(psds):  # convert to power for plotting
-            psds = (psds * psds.conj()).real
```
reverting, will not be needed.

drammock commented on mne/time_frequency/spectrum.py:
```diff
@@ -1226,11 +1197,21 @@ def __getitem__(self, item):
 
-def _check_data_shape(param, dim, data, expected):
-    if data.shape[dim] != expected:
+def _check_data_shape(data, freqs, info, ndim):
```
I reworked this to check everything at once rather than needing multiple calls to the check function. Seemed simpler this way once we knew we only had to handle 2D and 3D cases.

drammock commented on mne/time_frequency/tests/test_spectrum.py:
```diff
@@ -141,18 +142,21 @@ def test_spectrum_io(inst, tmp_path, request, evoked):
 
 
-def test_spectrum_copy(raw):
+def test_spectrum_copy(raw_spectrum):
```
Although I'd have preferred adding the fixture in its own PR, I decided it was expedient to just go with it, so I've used it in the other spectrum tests too.

drammock reviewed (approved):
here's a self-review as commentary on the changes I've just pushed.

@larsoner @agramfort or @mmagnuski maybe best for one of you to push the green button on this one, it's got a fair amount of code from both me and @alexrockhill

drammock commented on tutorials/simulation/10_array_objs.py:
```diff
@@ -224,40 +223,11 @@
 
-spectrum = spectrum_from_array(data=psd_ave, freqs=freqs, inst_info=info)
-spectrum.plot(picks=[0, 1], spatial_colors=False, exclude="bads")
+spectrum.plot()
```
```suggestion
spectrum.plot(spatial_colors=False)
```

drammock commented:
All green, this one is ready for review/merge

larsoner reviewed (approved):
Pretty short diff in the end, thanks @alexrockhill @drammock !
